### PR TITLE
WIP: AstroNvim v3.0.0

### PR DESCRIPTION
**Pull request by jay-babu**
Created so it is easier to leave comments. If you have another preference, please let me know

*NOTE:* This branch is not officially supported and we cannot guarantee everything works or is documented. If you want to try to use it please proceed with caution and know you will have to read the source code if things seem to not work as they did previously.

Development for v3.0.0 is basically finished and the WIP docs can be found here: https://v3.astronvim.pages.dev/nightly

### Tentative Release Cycle (Based on [Neovim v0.9 Release Cycle](https://github.com/neovim/neovim/milestone/28))

- March 1, 2023 : AstroNvim v3.0 release to nightly 
- March 31, 2023: AstroNvim v3.0 release to stable (Neovim v0.9 releases as stable)

### TODO

- [x] Update all existing documentation pages with new configuration structure
- [x] Update neo-tree buffer switching to match the rest of buffer switching
- [X] Move to dedicated theme plugin (https://github.com/AstroNvim/astrotheme)
- [x] Update [Default Plugin Configurations Page](https://v3.astronvim.pages.dev/nightly/Configuration/plugin_defaults)
- [x] Update documentation for the new `AstroNvim/astrotheme` plugin
- [x] Write the Migration to v3 Guide
- [x] Document installing AstroNvim on Neovim v0.8 with v2
- [x] Add documentation page for all AstroNvim exposed `User` `autocmd` events
- [x] Lazy load `astronvim` module and utility functions when needed (#1569)
- [x] Improve loading time of `core.mappings` (#1569)
- [x] Update documentation for #1569 changes
- [x] Validate utility restructuring
- [x] Add per buffer auto formatting control
- [x] Refactor fname functions in components to be functions of `self` rather than `bufnr` to get access to all the information in the Heirline component (similar to new `separated_path` component)
- [x] ~Fix bug where `neo-tree` closes when switching buffers, because of the new `mkview` stuff~ resolved in new neo-tree v2.50.2 update
- [ ] Add recipe for vim-tab based `tabline` with Heirline
- [ ] Resolve all v3 `TODO` notes from code

### CODE REVIEW

File: lua/core/mappings.lua [hunk 1]
```diff
@@ -15,15 +14,15 @@ maps.n["Q"] = "<Nop>"
...
--- Packer
-maps.n["<leader>pc"] = { "<cmd>PackerCompile<cr>", desc = "Packer Compile" }
-maps.n["<leader>pi"] = { "<cmd>PackerInstall<cr>", desc = "Packer Install" }
-maps.n["<leader>ps"] = { "<cmd>PackerSync<cr>", desc = "Packer Sync" }
-maps.n["<leader>pS"] = { "<cmd>PackerStatus<cr>", desc = "Packer Status" }
-maps.n["<leader>pu"] = { "<cmd>PackerUpdate<cr>", desc = "Packer Update" }
+-- Plugin Manager
+maps.n["<leader>pi"] = { function() require("lazy").install() end, desc = "Plugins Install" }
+maps.n["<leader>ps"] = { function() require("lazy").home() end, desc = "Plugins Status" }
+maps.n["<leader>pS"] = { function() require("lazy").sync() end, desc = "Plugins Sync" }
```
Comment (jay-babu) [inline lua/core/mappings.lua:20]: these seem to be swapped
```suggestion
maps.n["<leader>pS"] = { function() require("lazy").home() end, desc = "Plugins Status" }
maps.n["<leader>ps"] = { function() require("lazy").sync() end, desc = "Plugins Sync" }
```

Comment (mehalter) [reply lua/core/mappings.lua:20]: Yeah I realized the pattern we had before was the opposite of the convention of other lowercase/Capital bindings so that was a fix.

File: lua/core/mappings.lua [hunk 2]
```diff
@@ -252,6 +254,7 @@ if is_available "toggleterm.nvim" then
 end
 
 if is_available "nvim-dap" then
+  maps.n["<leader>D"] = sections.D
```
Comment (jay-babu) [inline lua/core/mappings.lua:257]: would it be possible to make these lowercase by default? discussed previously, I think we said we would do it in v3

PS: Happy New Year

Comment (mehalter) [reply lua/core/mappings.lua:257]: I'm not planning on changing this since `<leader>d` is a good binding for the dashboard and we have function key shortcuts for easy access for this functionality

Comment (mehalter) [reply lua/core/mappings.lua:257]: Hm we aren't binding `<leader>h` anymore, maybe we can make that like "Home" and do `<leader>d` for debugging

Comment (mehalter) [reply lua/core/mappings.lua]: Resolved in latest commits

**Comment by gwww** [comment]
WRT the rainbow plugin there are a couple of folks taking it over. Right now there's two repos with discussion of merging. https://www.reddit.com/r/neovim/comments/103x4ux/maintained_fork_of_nvimtsrainbow/

**Comment by fgimian** [comment]
Just wanted to report an issue with the current v3 branch on Windows using Neovim v0.8.2.  It seems that the configuration for lazy loading neo-tree is causing issues when the editor is launched multiple times with a directory as an argument.

e.g.

```powershell
nvim my-directory
```

The error is as follows:

```
[Neo-tree ERROR] debounce  hijack_netrw_1000  error:  ...data/lazy/neo-tree.nvim/lua/neo-tree/sources/manager.lua:39: attempt to index local 'state' (a nil value)
```

Switch lazy loading off for neo-tree resolves the issue, so I suspect the event being used may not be adequate.  Interestingly, I don't experience this problem inside WSL (i.e. on Linux).

Cheers
Fotis

**Comment by mehalter** [comment]
@fgimian could you try with the latest commit

**Comment by fgimian** [comment]
> @fgimian could you try with the latest commit

That did the trick, thank you so much! 😄

**Comment by fgimian** [comment]
Has the syntax for disabling a plugin changed in v3?  The following works with v2 but doesn't seem to have any affect in the v3 branch.

```lua
local config = {
  plugins = {
    init = {
      -- Disable better-escape for exiting insert mode with jj or jk.
      ["max397574/better-escape.nvim"] = { disable = true },
    },
  },
}

return config
```

Thanks heaps
Fotis

**Comment by mehalter** [comment]
@fgimian this notation has changed between packer and lazy. It should be `enabled = false` now

**Comment by mehalter** [comment]
Would definitely recommend reading lazy docs if you are going to try using v3

**Comment by fgimian** [comment]
> @fgimian this notation has changed between packer and lazy. It should be `enabled = false` now

Thanks, you beat me to the response 😄  It may be worth updating the example in https://github.com/AstroNvim/AstroNvim/blob/v3/lua/user_example/init.lua#L230

Really appreciate the fast reply!
Fotis

**Comment by fgimian** [comment]
> Would definitely recommend reading lazy docs if you are going to try using v3

Absolutely, I'm actually using lazy.nvim in my own config but didn't realise that AstroVim was just merging configs; I thought that maybe the `disabled` keyword was implemented by AstroVim itself.  I realised I was wrong when looking through the AstroVim source code. Sorry for the trouble.

Cheers
Fotis

**Comment by mehalter** [comment]
No problem! Also yeah documentation and stuff will be written/updated this month before v3 hits nightly on Jan 31 (TBD)

File: lua/plugins/mason.lua [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    default_config = function(opts)
+      local mason_lspconfig = require "mason-lspconfig"
+      mason_lspconfig.setup(opts)
+      mason_lspconfig.setup_handlers { function(server) astronvim.lsp.setup(server) end }
+      astronvim.event "LspSetup"
+    end,
+    dependencies = { "nvim-lspconfig" },
+  },
+  astronvim.plugin {
+    "jayp0521/mason-null-ls.nvim",
```
Comment (jay-babu) [inline lua/plugins/mason.lua:50]: `jay-babu`

File: lua/plugins/core.lua [hunk 4]
```diff
@@ -0,0 +1,131 @@
+return {
```
Comment (jay-babu) [inline lua/plugins/core.lua:1]: Thoughts on bringing `nvim-ts-rainbow` since there is a maintained fork?

Comment (jay-babu) [reply lua/plugins/core.lua:1]: https://github.com/mrjones2014/nvim-ts-rainbow

Comment (mehalter) [reply lua/plugins/core.lua:1]: @jay-babu has this been decided? There was discussion on whether or not this was going to move to a new organization as there were two people interested in maintaining and fully rewriting the plugin. Both currently have maintained forks and were going to discuss merging them. I will not be adding it until we can prove which fork will be maintained.

Comment (jay-babu) [reply lua/plugins/core.lua:1]: Oh, unsure.

https://www.reddit.com/r/neovim/comments/1079zgs/this_week_in_neovim_26_mon_jan_9th_2022/?utm_source=share&utm_medium=ios_app&utm_name=iossmf

Was simply going of this.

**Comment by willparsons** [comment]
👀

**Comment by hosaka** [comment]
As of https://github.com/rebelot/heirline.nvim/commit/7b57b27e9e4d1ffa4f3e149f8fdc927db18a850a a breaking change was introduced to `heirline.setup()`, it should now take a single setup argument. I'm attaching a patch for `configs/heirline.lua`, not sure if a PR would be more appropriate.

[heirline.patch](https://github.com/AstroNvim/AstroNvim/files/10399371/heirline.patch)

**Comment by mehalter** [comment]
> As of https://github.com/rebelot/heirline.nvim/commit/7b57b27e9e4d1ffa4f3e149f8fdc927db18a850a a breaking change was introduced to `heirline.setup()`, it should now take a single setup argument. I'm attaching a patch for `configs/heirline.lua`, not sure if a PR would be more appropriate.

This was fixed in this branch a couple days ago so no worries! That's for looking out!

**Comment by Mixaz017** [comment]
Looks like mappings for new file and telescope bookmarks have been changed, but alpha dashboard isn't updated accordingly. I think this needs to be fixed if mapping changes aren't temporary.

**Comment by mehalter** [comment]
@Mixaz017 thanks for the heads up! Fixed in the latest commit

**Comment by jay-babu** [comment]
Is there a plan to add `noice` as a plugin? Just wondering if I should add it myself or if `Astronvim` had plans to integrate with it

**Comment by mehalter** [comment]
We are not adding it @jay-babu

File: lua/core/utils/lsp.lua [hunk 5]
```diff
@@ -215,21 +209,28 @@ astronvim.lsp.capabilities.textDocument.completion.completionItem.tagSupport = {
...
 }
-astronvim.lsp.capabilities = user_plugin_opts("lsp.capabilities", astronvim.lsp.capabilities)
-astronvim.lsp.flags = user_plugin_opts "lsp.flags"
+astronvim.lsp.capabilities = user_opts("lsp.capabilities", astronvim.lsp.capabilities)
+astronvim.lsp.flags = user_opts "lsp.flags"
 
 --- Get the server settings for a given language server to be provided to the server's `setup()` call
 -- @param  server_name the name of the server
 -- @return the table of LSP options used when setting up the given language server
 function astronvim.lsp.server_settings(server_name)
```
Comment (jay-babu) [inline lua/core/utils/lsp.lua:218]: unsure, what your thoughts are on changing this to config since the directory changed as well?

```suggestion
function astronvim.lsp.config(server_name)
```

Comment (mehalter) [reply lua/core/utils/lsp.lua:218]: ah yeah that's a good idea

Comment (mehalter) [reply lua/core/utils/lsp.lua]: Changed in latest commit. Thanks for catching that @jay-babu !

**Comment by jay-babu** [comment]
```
:Notifications
2023-01-31T23:01:25 lazy.nvim  ERROR Failed to run `config` for heirline.nvim

...ocal/share/nvim/lazy/heirline.nvim/lua/heirline/init.lua:75: Vim:E355: Unknown option: statuscolumn

# stacktrace:
  - /heirline.nvim/lua/heirline/init.lua:75 _in_ **setup**
  - ~/.config/nvim/lua/plugins/configs/heirline.lua:111 _in_ **config**
```

unsure if this error is known. I keep getting this error. I tried pinning `heirline` to match the commit in the snapshot

https://gist.github.com/8586b9c99378c3f6d8d5f1c1c25de4cb

**Comment by mehalter** [comment]
> ```
> :Notifications
> 2023-01-31T23:01:25 lazy.nvim  ERROR Failed to run `config` for heirline.nvim
> 
> ...ocal/share/nvim/lazy/heirline.nvim/lua/heirline/init.lua:75: Vim:E355: Unknown option: statuscolumn
> 
> # stacktrace:
>   - /heirline.nvim/lua/heirline/init.lua:75 _in_ **setup**
>   - ~/.config/nvim/lua/plugins/configs/heirline.lua:111 _in_ **config**
> ```
> 
> unsure if this error is known. I keep getting this error. I tried pinning `heirline` to match the commit in the snapshot
> 
> https://gist.github.com/8586b9c99378c3f6d8d5f1c1c25de4cb
> 

Not sure what's going on with your config but that's not a problem. Could be you are trying to set it not in Neovim 0.9 so it's not available? I can confirm that this is working perfectly in AstroNvim v3 :)

**Comment by jay-babu** [comment]
Old version of neovim nightly, sorry. didn't realize, you are correct

File: lua/core/utils/init.lua [hunk 6]
```diff
@@ -282,37 +229,35 @@ end
 -- @param module the module path of the override setting
 -- @param default the default settings that will be overridden
 -- @param extend boolean value to either extend the default settings or overwrite them with the user settings entirely (default: true)
--- @param prefix a module prefix for where to search (default: user)
 -- @return the new configuration settings with the user overrides applied
-function astronvim.user_plugin_opts(module, default, extend, prefix)
+function astronvim.user_opts(module, default, extend)
   -- default to extend = true
```
Comment (jay-babu) [inline lua/core/utils/init.lua:234]: Is there any way you could keep `prefix` as an input parameter? I find this function really useful, if not, then I can copy it over

File: lua/core/mappings.lua [hunk 7]
```diff
@@ -300,18 +338,19 @@ end
...
+maps.n["<leader>us"] = { function() astronvim.ui.toggle_spell() end, desc = "Toggle spellcheck" }
+maps.n["<leader>uS"] = { function() astronvim.ui.toggle_conceal() end, desc = "Toggle conceal" }
 maps.n["<leader>ut"] = { function() astronvim.ui.toggle_tabline() end, desc = "Toggle tabline" }
 maps.n["<leader>uu"] = { function() astronvim.ui.toggle_url_match() end, desc = "Toggle URL highlight" }
 maps.n["<leader>uw"] = { function() astronvim.ui.toggle_wrap() end, desc = "Toggle wrap" }
 maps.n["<leader>uy"] = { function() astronvim.ui.toggle_syntax() end, desc = "Toggle syntax highlight" }
-maps.n["<leader>uN"] = { function() astronvim.ui.toggle_ui_notifications() end, desc = "Toggle UI notifications" }
 
-astronvim.set_mappings(astronvim.user_plugin_opts("mappings", maps))
+astronvim.set_mappings(astronvim.user_opts("mappings", maps))
```
Comment (jay-babu) [inline lua/core/mappings.lua:356]: I am not for/against for `mappings.lua`, is there a reason to not use `lazy.nvim`, https://github.com/folke/lazy.nvim#:~:text=load%20on%20filetype-,keys,-string%3F%20or?

Comment (mehalter) [reply lua/core/mappings.lua:356]: There are a lot of reasons here mainly regarding ease of control to the user. They don't need to spend time searching through source code to figure out which plugin binds a mapping, if they don't want one they can set it to false in one table. Also it keeps the user from safely overriding mappings in a methodological way. If the user defines a mapping in the mappings table you know it will only be set once and will be safe. With lazy it will be down to what plugin is loaded first. There are a lot more reasons for this as well, but those are a few.

TL;DR this increases usability, ease of configuration to the user, stability, and safety by an extremely large margin

File: lua/core/utils/init.lua (hunk 6 above)
Comment (mehalter) [reply lua/core/utils/init.lua:234]: Sure, what do you use the function for? I could make it so the prefix is default not `user` maybe. How are you using it in your configuration?

Comment (mehalter) [reply lua/core/utils/init.lua:234]: @jay-babu I was looking through it and I'm not sure how this function is useful with the prefix? All it does is change whether or not it's in the `user` module. And in your user configuration you should always be getting from the `user` module so I don't think the prefix is necessary. You can just add your prefix to the `module` parameter

Comment (jay-babu) [reply lua/core/utils/init.lua:234]: That makes sense. I will attempt to do that

File: lua/core/lazy.lua [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+vim.opt.rtp:prepend(lazypath)
+
+local spec = astronvim.updater.options.pin_plugins and { { import = astronvim.updater.snapshot.module } } or {}
+
+local user_plugins = astronvim.user_opts "plugins"
+for _, config_dir in ipairs(astronvim.supported_configs) do
+  if vim.fn.isdirectory(config_dir .. "/lua/user/plugins") == 1 then user_plugins = { import = "user.plugins" } end
+end
+
+vim.list_extend(spec, { { import = "plugins" }, user_plugins })
```
Comment (jay-babu) [inline lua/core/lazy.lua:28]: I tried playing with this, but was unable to do so successfully. From what I understand `user.plugins` would be imported into `lazy.nvim`. I was wanted to know if there was a way to specify multiple places to import? like add on `user.custom.plugins` to lazy.nvim import

Comment (mehalter) [reply lua/core/lazy.lua:28]: You can do something like this in your `user/lazy.lua` (or in the correct place in your `user/init.lua`):

```lua
return function(config)
  vim.list_extend(config.spec, { { import = "user.custom.plugins" } })
  return config
end
```

Comment (jay-babu) [reply lua/core/lazy.lua:28]: Thanks!

**Comment by gwww** [comment]
The refactor broke starting lazygit because `M.user_terminals` is nil.

```
E5108: Error executing lua: /Users/glenn/.config/nvim/lua/core/utils/init.lua:97: attempt to index local 'terms' (a nil value)
stack traceback:
        /Users/glenn/.config/nvim/lua/core/utils/init.lua:97: in function 'toggle_term_cmd'
        /Users/glenn/.config/nvim/lua/core/mappings.lua:269: in function </Users/glenn/.config/nvim/lua/core/mappings.lua:269>
```

**Comment by khaneliman** [comment]
> The refactor broke starting lazygit because `M.user_terminals` is nil.
> 
> ```
> E5108: Error executing lua: /Users/glenn/.config/nvim/lua/core/utils/init.lua:97: attempt to index local 'terms' (a nil value)
> stack traceback:
>         /Users/glenn/.config/nvim/lua/core/utils/init.lua:97: in function 'toggle_term_cmd'
>         /Users/glenn/.config/nvim/lua/core/mappings.lua:269: in function </Users/glenn/.config/nvim/lua/core/mappings.lua:269>
> ```

Yeah, been trying to figure out how to fix that one too with that latest commit that did it

**Comment by gwww** [comment]
A work around fix that worked for me was to add the line `M.user_terminals = {}` to `core/utils/init.lua`. It makes things work for me but I don't know if its the right long term fix. But its GEFN!

**Comment by mehalter** [comment]
Cool I'll fix this next week @gwww thanks for reporting!

**Comment by khaneliman** [comment]
Yeah, looks like user_terminals is declared in bootstrap.lua and it's referencing M.user_terminals in init.lua

**Comment by mehalter** [comment]
@gwww see if the latest commit fixes it. No idea if it works 😂 don't have a computer on me to test it and made the commit from my phone 😅

**Comment by gwww** [comment]
Works!

**Comment by jay-babu** [comment]
just fyi, `jay-babu` plugins have onboarded to semver and git tagging, so no need to pin them anymore

**Comment by mehalter** [comment]
> just fyi, `jay-babu` plugins have onboarded to semver and git tagging, so no need to pin them anymore

That's great! I just updated the snapshot to reflect this!

**Comment by khaneliman** [comment]
Struggling to find how to properly add new cmp sources in v3. Do you remember what commit changed it or where the code I should look at that would help is? 

<details>
  <summary>cmp.lua</summary>

```lua
{
		"hrsh7th/cmp-calc",
		dependencies = {
			"hrsh7th/nvim-cmp",
		},
		event = "InsertEnter",
		config = function()
			local utils = require("core.utils")
			-- local cmp = require("cmp")
			-- local config = cmp.get_config()
			-- table.insert(config.sources, {
			-- 	{ name = "calc" },
			-- })
			-- cmp.setup(config)
			utils.add_user_cmp_source("calc")
		end,
	},
	{
		"hrsh7th/cmp-emoji",
		dependencies = {
			"hrsh7th/nvim-cmp",
		},
		event = "InsertEnter",
		config = function()
			local cmp = require("cmp")
			local config = cmp.get_config()
			table.insert(config.sources, {
				{ name = "emoji" },
			})
			cmp.setup(config)
		end,
	},
```
</details>

**Comment by mehalter** [comment]
> Struggling to find how to properly add new cmp sources in v3. Do you remember what commit changed it or where the code I should look at that would help is? I found a hacky workaround to just overwrite the entire source list... but it feels wrong. 
> 
> <details>
>   <summary>nvim/lua/user/plugins/cmp.lua</summary>
> 
> ```lua
> {
> 		"hrsh7th/cmp-calc",
> 		dependencies = {
> 			"hrsh7th/nvim-cmp",
> 		},
> 		event = "InsertEnter",
> 		config = function()
> 			local utils = require("core.utils")
> 			-- local cmp = require("cmp")
> 			-- local config = cmp.get_config()
> 			-- table.insert(config.sources, {
> 			-- 	{ name = "calc" },
> 			-- })
> 			-- cmp.setup(config)
> 			utils.add_user_cmp_source("calc")
> 		end,
> 	},
> 	{
> 		"hrsh7th/cmp-emoji",
> 		dependencies = {
> 			"hrsh7th/nvim-cmp",
> 		},
> 		event = "InsertEnter",
> 		config = function()
> 			local cmp = require("cmp")
> 			local config = cmp.get_config()
> 			table.insert(config.sources, {
> 				{ name = "emoji" },
> 			})
> 			cmp.setup(config)
> 		end,
> 	},
> ```
> </details>
> 
> <details>
> <summary>possible workaround</summary>
> 
> ```lua
> return {
> 	{
> 		"hrsh7th/nvim-cmp",
> 		opts = function(_, opts)
> 			-- require("plugins.cmp")(plugin, opts) -- include the default astronvim config that calls the setup call
> 			opts.sources = {
> 				opts.sources,
> 				{ name = "nvim_lsp", priority = 1000 },
> 				{ name = "omni",     priority = 800 },
> 				{ name = "luasnip",  priority = 750 },
> 				{ name = "emoji",    priority = 700 },
> 				{ name = "calc",     priority = 650 },
> 				{ name = "path",     priority = 500 },
> 				{ name = "fish",     priority = 300 },
> 				{ name = "npm",      priority = 300 },
> 				{ name = "git",      priority = 300 },
> 				{ name = "buffer",   priority = 250 },
> 			}
> 			return opts
> 		end,
> 	},
> 	{
> 		"hrsh7th/cmp-calc",
> 		dependencies = {
> 			"hrsh7th/nvim-cmp",
> 		},
> 		event = "InsertEnter",
> 		-- config = function()
> 		-- 	local cmp = require("cmp")
> 		-- 	local config = cmp.get_config()
> 		-- 	table.insert(config.sources, {
> 		-- 		{ name = "calc" },
> 		-- 	})
> 		-- 	cmp.setup(config)
> 		-- end,
> 	},
> 
> ```
> 
> </details>

Check out this new development deployment of the documentation: https://v3.astronvim.pages.dev/nightly/Recipes/cmp#add-more-sources

**Comment by khaneliman** [comment]
> > Struggling to find how to properly add new cmp sources in v3. Do you remember what commit changed it or where the code I should look at that would help is? I found a hacky workaround to just overwrite the entire source list... but it feels wrong.
> > 
> >   nvim/lua/user/plugins/cmp.lua
> > ```lua
> > {
> > 		"hrsh7th/cmp-calc",
> > 		dependencies = {
> > 			"hrsh7th/nvim-cmp",
> > 		},
> > 		event = "InsertEnter",
> > 		config = function()
> > 			local utils = require("core.utils")
> > 			-- local cmp = require("cmp")
> > 			-- local config = cmp.get_config()
> > 			-- table.insert(config.sources, {
> > 			-- 	{ name = "calc" },
> > 			-- })
> > 			-- cmp.setup(config)
> > 			utils.add_user_cmp_source("calc")
> > 		end,
> > 	},
> > 	{
> > 		"hrsh7th/cmp-emoji",
> > 		dependencies = {
> > 			"hrsh7th/nvim-cmp",
> > 		},
> > 		event = "InsertEnter",
> > 		config = function()
> > 			local cmp = require("cmp")
> > 			local config = cmp.get_config()
> > 			table.insert(config.sources, {
> > 				{ name = "emoji" },
> > 			})
> > 			cmp.setup(config)
> > 		end,
> > 	},
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 
> > possible workaround
> > ```lua
> > return {
> > 	{
> > 		"hrsh7th/nvim-cmp",
> > 		opts = function(_, opts)
> > 			-- require("plugins.cmp")(plugin, opts) -- include the default astronvim config that calls the setup call
> > 			opts.sources = {
> > 				opts.sources,
> > 				{ name = "nvim_lsp", priority = 1000 },
> > 				{ name = "omni",     priority = 800 },
> > 				{ name = "luasnip",  priority = 750 },
> > 				{ name = "emoji",    priority = 700 },
> > 				{ name = "calc",     priority = 650 },
> > 				{ name = "path",     priority = 500 },
> > 				{ name = "fish",     priority = 300 },
> > 				{ name = "npm",      priority = 300 },
> > 				{ name = "git",      priority = 300 },
> > 				{ name = "buffer",   priority = 250 },
> > 			}
> > 			return opts
> > 		end,
> > 	},
> > 	{
> > 		"hrsh7th/cmp-calc",
> > 		dependencies = {
> > 			"hrsh7th/nvim-cmp",
> > 		},
> > 		event = "InsertEnter",
> > 		-- config = function()
> > 		-- 	local cmp = require("cmp")
> > 		-- 	local config = cmp.get_config()
> > 		-- 	table.insert(config.sources, {
> > 		-- 		{ name = "calc" },
> > 		-- 	})
> > 		-- 	cmp.setup(config)
> > 		-- end,
> > 	},
> > ```
> 
> Check out this new development deployment of the documentation: https://v3.astronvim.pages.dev/nightly/Recipes/cmp#add-more-sources

Oh you do have a v3 wip documentation out there. Thanks, I'll browse through at instead of the nightly stuff. I swear I tried that way, but i'll try it again. thanks

**Comment by gwww** [comment]
Did you consider checking if the core plugins are all installed as a part of check health?
